epstein/src/main/scala/epstein/ZQuery.scala

```diff
@@ -0,0 +1,211 @@
+    }
+
+  /**
+   * Returns an effect that models executing this query, returning the query
+   * result along with the cache containing a complete log of all requests
+   * executed and their results. This can be useful for logging or analysis of
+   * query execution.
+   */
+  final def runLog: ZIO[R, E, (Cache, A)] =
+    for {
+      cache <- Cache.empty
+      a     <- runCache(cache)
+    } yield (cache, a)
+
+  /**
+   * Returns a query that models the execution of this query and the specified
+   * query in parallel, combining their results into a tuple. All requests that
+   * can potentially be executed in parallel will be batched and deduplication
+   * and caching will be applied.
+   */
+  final def zipPar[R1 <: R, E1 >: E, B](that: ZQuery[R1, E1, B]): ZQuery[R1, E1, (A, B)] =
+    zipWithPar(that)((_, _))
+
+  /**
+   * Returns a query that models the execution of this query and the specified
+   * query in parallel, combining their results with the specified function.
+   * All requests that can potentially be executed in parallel will be batched
+   * and deduplication and caching will be applied.
+   */
+  final def zipWithPar[R1 <: R, E1 >: E, B, C](that: ZQuery[R1, E1, B])(f: (A, B) => C): ZQuery[R1, E1, C] =
+    new ZQuery[R1, E1, C] {
+      def step(cache: Cache): ZIO[R1, E1, Result[R1, E1, C]] =
+        self.step(cache).zip(that.step(cache)).map {
+          case (Result.Blocked(br1, c1), Result.Blocked(br2, c2)) => Result.blocked(br1 ++ br2, c1.zipWithPar(c2)(f))
+          case (Result.Blocked(br, c), Result.Done(_))            => Result.blocked(br, c.zipWithPar(that)(f))
+          case (Result.Done(_), Result.Blocked(br, c))            => Result.blocked(br, self.zipWithPar(c)(f))
+          case (Result.Done(a), Result.Done(b))                   => Result.done(f(a, b))
+        }
+    }
+}
+
+object ZQuery {
+
+  /**
+   * Collects a collection of queries into a query returning a collection of
+   * their results. All requests will be batched and deduplication and caching
+   * will be applied.
+   */
+  final def collectAllPar[R, E, A](as: Iterable[ZQuery[R, E, A]]): ZQuery[R, E, List[A]] =
+    foreachPar(as)(identity)
+
+  /**
+   * Constructs a query that fails with the specified error.
+   */
+  final def fail[E](error: E): ZQuery[Any, E, Nothing] =
+    ZQuery(ZIO.fail(error))
+
+  /**
+   * Performs a query for each element in a collection, collecting the results
+   * into a query returning a collection of their results. All requests will be
+   * batched and caching and deduplication will be applied.
+   */
+  final def foreachPar[R, E, A, B](as: Iterable[A])(f: A => ZQuery[R, E, B]): ZQuery[R, E, List[B]] =
+    as.foldRight[ZQuery[R, E, List[B]]](ZQuery.succeed(Nil))((a, bs) => f(a).zipWithPar(bs)(_ :: _))
+
+  /**
+   * Constructs a query from an effect.
+   */
+  final def fromEffect[R, E, A](effect: ZIO[R, E, A]): ZQuery[R, E, A] =
+    ZQuery(effect.map(Result.done))
```

Comment:
Could it be useful to have `provide` directly on `ZQuery` to avoid passing it at each run?